feat(compiler-plugin): add two simple checkers on the frontend part

FreshMag opened the pull request:
Started working on the frontend part of the compiler plugin.

Implemented two checkers:
- `NoAlignInsideALoop`: catches aggregate operations inside a loop without manual `alignedOn`
- `NoAlignOrDealign`: catches explicit usage of `align` and `dealign` inside aggregate functions.

DanySK commented on gradle-plugin/src/main/kotlin/it/unibo/.DS_Store:
Please remove this one from the tracker

DanySK commented on settings.gradle.kts:
```diff
@@ -32,4 +32,4 @@ gitHooks {
 
 includeBuild("compiler-plugin")
-include("alchemist-incarnation-collektive", "compiler-embeddable", "dsl", "stdlib")
+include("alchemist-incarnation-collektive", "compiler-embeddable", "dsl", "stdlib", "compiler-plugin-test")
```
sort alphabetically and end the file with a newline

DanySK commented on compiler-plugin-test/src/test/resources/TestAggregateInLoop.kt:
```diff
@@ -0,0 +1,13 @@
+        %s
+    }
+    %s
+}
```
Can you comment this code? It is unclear what it does

DanySK commented on compiler-plugin-test/src/test/resources/SingleAggregateLine.kt:
```diff
@@ -0,0 +1,5 @@
+
+fun Aggregate<Int>.x() {
+    %s
+}
```
Can you comment the meaning of %s?

DanySK commented on compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/TestLoopWithoutAlignWarning.kt:
```diff
@@ -0,0 +1,41 @@
+
+@OptIn(ExperimentalCompilerApi::class)
+class TestLoopWithoutAlignWarning : FreeSpec({
+    val expectedWarningMessage = "Warning: aggregate function \"exampleAggregate\" called inside a loop " +
```
`const` in the Companion?

DanySK commented on compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/MissingAlignExtension.kt:
```diff
@@ -0,0 +1,108 @@
+        override val functionCallCheckers: Set<FirFunctionCallChecker>
+            get() = setOf(NoAlignInsideALoop, NoAlignOrDealign)
+    }
+}
```
Unless all top level members can be private but one, I would split this into multiple files, possibly creating a new subpackage.

DanySK reviewed (changes requested):
reviewed 6/13, left comments to address

angelacorte commented on compiler-plugin-test/src/test/resources/TestAggregateInLoop.kt:
```diff
@@ -0,0 +1,13 @@
+import it.unibo.collektive.aggregate.api.Aggregate
+
+fun Aggregate<Int>.exampleAggregate(): Int = 0
```
Can I suggest adding a test where you use the aggregate constructs (neighboring or exchange etc)? In order to have a more complete test.

FreshMag replied:
Ok, I've added a more flexible unit test in [161237f](https://github.com/Collektive/collektive/pull/454/commits/161237f864d9d37dc4846fd281ce243a8e3d21c1).
I've added a parametric line of code where it is possible to specify the aggregate function to put inside the loop.

angelacorte commented on gradle/libs.versions.toml:
```diff
@@ -27,6 +26,7 @@ kotlin-test-annotations-common = { module = "org.jetbrains.kotlin:kotlin-test-an
 kotlin-gradle-plugin = { module = "org.jetbrains.kotlin:kotlin-gradle-plugin", version.ref = "kotlin" }
 kotlinpoet = "com.squareup:kotlinpoet:1.18.1"
 kotlinx-coroutines = { module = "org.jetbrains.kotlinx:kotlinx-coroutines-core", version.ref = "coroutines" }
+kctfork = "dev.zacsweers.kctfork:core:0.5.1"
```
can I suggest to follow the overall semantics? 
for example from this
``` kctfork = "dev.zacsweers.kctfork:core:0.5.1" ```
to this
``` kctfork = { module = "dev.zacsweers.kctfork:core", version = "0.5.1" } ```

DanySK replied:
Why? It is just longer and we gain nothing. Unless we use `version.ref`, I'd keep the short version.

The only suggestion here is to use the lexicographic order to simplify the file reading.

FreshMag replied:
I've left the previous format as @DanySK suggested, for now. Also, I've adopted lexicographic order.

DanySK commented:
@FreshMag you did not allow modifications by maintainers, hence we cannot automatically rebase or merge the mainline into your branch. You need to keep it up to date yourself. For the next PRs, I recommend allowing them.

DanySK commented:
@FreshMag make sure all conversations are solved before asking for another review round :smile:

nicolasfara commented on compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/TestLoopWithoutAlignWarning.kt:
```diff
@@ -0,0 +1,62 @@
+        listOf(
+            "exampleAggregate" to "exampleAggregate()",
+            "neighboring" to "neighboring(0)",
+        ).forEach {
```
I prefer tuple-unpacking for more clear code.

```kotlin
.forEach { (functionName, functionCall) -> .... }
```

DanySK replied:
it's called "destructuring declaration" in Kotlin jargon, but I agree.

nicolasfara commented on compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/TestExplicitAlign.kt:
```diff
@@ -0,0 +1,29 @@
+class TestExplicitAlign : FreeSpec({
+    val testingProgramTemplate = CompileUtils.testingProgramFromTemplate(ProgramTemplates.SINGLE_AGGREGATE_LINE)
+
+    "An explicit align function should not be used" - {
```
Please, rewrite the test names to be more "flowable" when you read it (nested test assertions).

For instance, I would change as follows:

```kotlin
"The `align` function" - {
  "should produce a warning when used explicitly" {
  ...
}
```

Similarly for the rest of the test assertions.

nicolasfara commented on compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/TestLoopWithoutAlignWarning.kt:
```diff
@@ -0,0 +1,62 @@
+
+@OptIn(ExperimentalCompilerApi::class)
+class TestLoopWithoutAlignWarning : FreeSpec({
+    "A single aggregate function called inside a loop: " - {
```
Also here make the test assertions more flowable.

nicolasfara commented on compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/checkers/CheckersUtility.kt:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    /**
+     * Checks is a specific receiver parameter is `Aggregate` (`Aggregate<ID>.example()`).
+     * @param session [FirSession] session of the [CheckerContext].
+     * @return [Boolean] **true** if it's an `Aggregate` function, **false** otherwise.
```
Use markdown syntax to document functions.

https://kotlinlang.org/docs/kotlin-doc.html#inline-markup

nicolasfara commented on compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/checkers/CheckersUtility.kt:
```diff
@@ -0,0 +1,57 @@
+     * @return [Boolean] **true** if it's an `Aggregate` function, **false** otherwise.
+     */
+    fun FirReceiverParameter.isAggregate(session: FirSession): Boolean =
+        this.typeRef.toClassLikeSymbol(session)?.name?.asString() == "Aggregate"
```
Remove `this`

nicolasfara commented on compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/checkers/ExplicitAlignDealign.kt:
```diff
@@ -0,0 +1,27 @@
+            reporter.reportOn(
+                expression.calleeReference.source,
+                PluginErrors.DOT_CALL_WARNING,
+                "Warning: \"%s\" method should not be explicitly used".format(calleeName),
```
```suggestion
                "Warning: '$calleeName' method should not be explicitly used",
```

Generally, try to avoid `format` since you have string template

nicolasfara commented:
I have two general considerations for this PR:

- The provided tests are not multiplatform, but JVM only (easy fixable I guess)
- From what I can understand, we test that an aggregate function inside a loop without `alignOn` should produce a warning. But what about `listOf(1, 2, 3, 4, 5).forEach { neighboring(it) }`? The same for `map`, `filter`, and so on... at the current stage, do we want to intercept also these cases?

@DanySK @angelacorte

DanySK commented:
1. Better fix the MP part before merging. This PR is a kickstart and should build a solid basis for the next steps.
2. I'd go one case at a time. Let's do the for loop first, then we will improve step by step

FreshMag commented:
Alright, I have changed the requested parts. Have you got any advice on how to fix the MP part?

nicolasfara commented:
> Does the library you're using support compilation for KMP? Otherwise I think we'll only test the JVM side of the compiler.

Taking a brief look at the dependencies used in this module, for sure
```
apache-commons-codec = "commons-codec:commons-codec:1.17.1"
apache-commons-text = "org.apache.commons:commons-text:1.12.0"
```

those two dependencies are JVM only...

If we want to support KMP we should find an alternative to those libraries.

FreshMag commented:
The first library is not used, probably a typo by me. I will fix it within the next few hours (by the way, it is also present inside the `compiler-embeddable` project).

The second library is used just for string utilities, but it can be easily replaced by implementing a simple method for string substitution. 

I'm not sure if the [Kotlin compile-testing library](https://github.com/ZacSweers/kotlin-compile-testing) can be a problem as well?

DanySK commented on compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/util/CompileUtils.kt:
```diff
@@ -0,0 +1,93 @@
+        }.compile()
+    }
+
+    data class KotlinTestingProgram(
+        val fileName: String,
+        private val template: String,
+        val program: String,
+        private val properties: Map<String, String>,
+    ) {
+
+        fun formatCode(vararg args: Any?): KotlinTestingProgram =
+            KotlinTestingProgram(fileName, template, program.format(*args), properties)
+
+        fun put(key: String, value: String): KotlinTestingProgram {
+            val updateProperties = properties + (key to value)
+            return KotlinTestingProgram(
+                fileName,
+                template,
+                StringSubstitutor.replace(template, updateProperties, "%(", ")"),
+                updateProperties,
+            )
+        }
+
+        fun import(javaClass: Class<*>): KotlinTestingProgram =
+            put(
+                "imports",
+                properties["imports"] + "\nimport " + javaClass.name,
+            )
+
+        infix fun shouldCompileWith(compilationCheck: (JvmCompilationResult) -> Unit) {
+            println(program)
+            println(properties)
+            val result = compile(fileName, program)
+            result.exitCode shouldBe KotlinCompilation.ExitCode.OK
+            compilationCheck(result)
+        }
+    }
```
@FreshMag do you prefer to merge this first and then try to build the DSL (maybe with @nicolasfara's help) or do you want to try it before merging?

DanySK commented:
@nicolasfara if they're only used in the test project, and if the test library is jvm-only, then it's not a problem is it?

nicolasfara commented:
> @nicolasfara if they're only used in the test project, and if the test library is jvm-only, then it's not a problem is it?

No, it is not a problem: since this test project is meant to test aggregate anti-pattern or problematic code, it is not mandatory to make it KMP. At the end of the day, at this layer (frontend plugin) we should not deal with platform-specific stuff.

nicolasfara commented on compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/util/CompileUtils.kt:
```diff
@@ -0,0 +1,93 @@
+            )
+
+        infix fun shouldCompileWith(compilationCheck: (JvmCompilationResult) -> Unit) {
+            println(program)
+            println(properties)
```
Are all these `println` needed or are they something from your experiments?

FreshMag replied:
I'm removing them right now, sorry. They are not needed.

FreshMag commented:
I have removed the `apache` libraries that were included inside the testing project. I will leave the Kotlin compile testing library as suggested.

As a first attempt, I have implemented a possible, incomplete DSL to replace the current program template formatting methodology. You can examine it in [this draft](https://github.com/FreshMag/collektive-plugin/pull/6). 
Let me know what do you think about it and please provide possible improvements if needed.

nicolasfara commented on compiler-plugin-test/build.gradle.kts:
Just realised that this module should not be published in Maven Central, right? @DanySK 

To disable the module's publication the following lines should be added:

```
tasks {
    // Prevent publishing this module since it is a test-only module
    withType<AbstractPublishToMaven>().configureEach {
        enabled = false
    }
    withType<GenerateModuleMetadata>().configureEach {
        enabled = false
    }
}
```

DanySK replied:
yeah, makes sense

nicolasfara commented on compiler-plugin-test/src/test/resources/SingleAggregateInLoop.template.kt:
```diff
@@ -0,0 +1,20 @@
+    The parts after "%" are placeholders for string formatting that allow putting an arbitrary
+    piece of code inside them (avoiding having to create multiple testing sources).
+*/
+fun Aggregate<Int>.x(pivot: (Int) -> Any?) {
```
Why do you specify the `pivot` as the input argument?

nicolasfara replied:
In that case, the `localId` divides devices based on their ID (even and odd).
In this case, I would avoid having the pivot based on the `localId`

FreshMag replied:
Ok, will fix it in a moment.

nicolasfara commented on compiler-plugin-test/src/test/kotlin/it/unibo/collektive/test/LoopWithoutAlignTest.kt:
```diff
@@ -0,0 +1,62 @@
+            }
+            "using $functionName wrapped in a specific alignedOn" - {
+                val testingProgram = testingProgramTemplate
+                    .put("beforeAggregate", "alignedOn(pivot(localId)) {")
```
I'm pretty sure that creating a pivot based on `localId` can be problematic in real code: since all the devices have different IDs, no one would align at the same path (pivot) since it is built based on the ID.

Since here we are testing whether the warning is raised or not, but since the tests act (among the many) as a specification, I would not make the pivot dependent on the `localId`, what do you think @angelacorte @DanySK ?

DanySK replied:
For testing purposes, I don't really see the problem

nicolasfara replied:
I would have preferred a static string instead, but that's okay.
Mark as resolved.

nicolasfara commented:
> I have removed the `apache` libraries that were included inside the testing project. I will leave the Kotlin compile testing library as suggested.
> 
> As a first attempt, I have implemented a possible, incomplete DSL to replace the current program template formatting methodology. You can examine it in [this draft](https://github.com/FreshMag/collektive-plugin/pull/6). Let me know what do you think about it and please provide possible improvements if needed.

I'll take a look in a moment

nicolasfara commented on compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/checkers/ExplicitAlignDealign.kt:
```diff
@@ -0,0 +1,27 @@
+            reporter.reportOn(
+                expression.calleeReference.source,
+                PluginErrors.DOT_CALL_WARNING,
+                "Warning: \"$calleeName\" method should not be explicitly used",
```
```suggestion
                "Warning: '$calleeName' method should not be explicitly used",
```
